src/services/componentService.ts

```diff
@@ -201,6 +209,75 @@ export const fetchAndStoreComponent = async (
   }
 };
 
+/**
+ * Fetch and store a single component by spec, text, or URL (in that order).
+ * Returns the populated ComponentSpec or null.
+ */
+export const fetchAndStoreComponent = async (
+  component: ComponentReference,
+): Promise<ComponentSpec | null> => {
+  try {
+    const id = `component-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
+    const createdAt = Date.now();
+    const updatedAt = Date.now();
+
+    let text: string | undefined;
+    let spec: ComponentSpec | null | undefined;
+
+    // 1. If spec is present, use it and serialize to text
+    if (component.spec) {
+      spec = component.spec;
+      text = component.text ?? yaml.dump(component.spec);
+      if (text && component.url) {
+        await saveComponent({
+          id,
+          url: component.url,
+          data: text,
+          createdAt,
+          updatedAt,
+        });
+      } else {
+        console.error("Unable to save component loaded from spec without URL");
```

Comment:
>component loaded from spec

This error seems incorrect since you can get here even without doing "spec -> text".